ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/Code.java

```diff
@@ -100,6 +104,35 @@ public boolean isValidJumpDestination(
     return validJumpDestinations.get(jumpDestination);
   }
 
+  /**
+   * Determine whether a specified destination is a valid jump sub target.
+   *
+   * @param evm the EVM executing this code
+   * @param frame The current message frame
+   * @param destination The destination we're checking for validity.
+   * @return Whether or not this location is a valid jump sub destination.
+   */
+  public boolean isValidJumpSubDestination(
+      final EVM evm, final MessageFrame frame, final UInt256 destination) {
+    if (!destination.fitsInt()) return false;
+    final int jumpDestination = destination.intValue();
+    if (jumpDestination > getSize()) return false;
+
+    if (validJumpSubDestinations == null) {
+      // Calculate valid jump destinations
+      validJumpSubDestinations = new BitSet(getSize());
+      evm.forEachOperation(
+          this,
+          frame.getContractAccountVersion(),
+          (final Operation op, final Integer offset) -> {
+            if (op.getOpcode() == BeginSubOperation.OPCODE) {
+              validJumpSubDestinations.set(offset);
+            }
+          });
+    }
+    return validJumpSubDestinations.get(jumpDestination);
+  }
```

Comment:
I see that this is modeled after the existing `isValidJumpDestination`. What do you think about moving the caching logic for both to the fields themselves using `Suppliers.memoize`? I think it'd make it easier to read and getting rid of a null in the code makes angels sing :)

Reply:
If I understood your proposal. One way to do that is to create suppliers in the constructor of the `Code` class. This requires me to add the `EVM` and the "contractAccountVersion" to the constructor and to impact many classes. I wonder if it might not be too impactful

```java
 public Code(final EVM evm, final int contractAccountVersion, final Bytes bytes) {
    this.bytes = bytes;

    validJumpDestinations = Suppliers.memoize(()-> {
      final BitSet bitSet = new BitSet(getSize());
      evm.forEachOperation(
              this,
              contractAccountVersion,
              (final Operation op, final Integer offset) -> {
                if (op.getOpcode() == JumpDestOperation.OPCODE) {
                  bitSet.set(offset);
                }
              });
      return bitSet;
    });

    validJumpSubDestinations = Suppliers.memoize(()-> {
      final BitSet bitSet = new BitSet(getSize());
      evm.forEachOperation(
              this,
              contractAccountVersion,
              (final Operation op, final Integer offset) -> {
                if (op.getOpcode() == BeginSubOperation.OPCODE) {
                  bitSet.set(offset);
                }
              });
      return bitSet;
    });
  }
```

Reply:
I think you're right, this would be too impactful. Looks good as is then :)